lib/compile/logger.js

```diff
@@ -0,0 +1,41 @@
+'use strict';
+
+module.exports = Logger;
+
+/**
+ * @constructor
+ * @this Logger
+ * @param {*=} loggerObj logger object, expected to implement log, warn and error, if not given global console will be taken.
+ */
+function Logger(loggerObj) {
+  if(loggerObj && typeof loggerObj === 'object')
+    this.logger = loggerObj;
+  else if(typeof loggerObj === 'undefined')
+    this.logger = typeof console === 'object' ? console : {log: noop, warn: noop, error: noop};
+  else
+    this.logger = null;
+}
+
+Logger.prototype.log = function() {
+  _logFunc.call(this, 'log', arguments);
+};
+
+Logger.prototype.warn = function() {
+  _logFunc.call(this, 'warn', arguments);
+};
+
+Logger.prototype.error = function() {
+  _logFunc.call(this, 'error', arguments);
+};
+
+/**
+ * @this Logger
+ * @param {String} level log level - log, warn or error
+ * @param {Array} args log argumets
+ * @private
+ */
+function _logFunc(level, args) {
+  if(this.logger && this.logger[level]) this.logger[level].apply(this, args);
+}
+
+function noop() {}
```

Comment:
This whole Logger class seems like an unnecessary indirection, to be honest.
It can be much simpler without it completely:
1. if undefined is passed - use console (in Ajv constructor)
2. if false is passed assign a noop object: `{log: noop, warn: noop, error: noop}`
3. otherwise just use what is passed

Reply:
It seems to me that it is more safe to use the Logger class because:
1. it is more testable - I was able to write a standalone tests for all the logger scenarios (logger.spec.js), if I move the logic into Ajv ctor it will be much harder as I will have to find specific scenarios with Ajv that leads to every console call (log, warn, error) - and anyhow it might change in the future
2. partial custom logger implementation can leads to an error - this code checks for the log function existence before calling it

Reply:
1. You won't have to test it if it doesn't exist. You only really need to test 3 scenarios for the option: logs to console/log to passed logger/doesn't log.
2. It essentially hides the method in passed logger. Throwing an exception is better. If a user want's to hide say warnings but not errors they just need to pass logger with a noop for a method.

Reply:
1. I see what you mean - still we have to keep in mind that the test is a bit implied as we test a scenario that we knows that currently happen to write to the log - the test does not control it so it is less "unit" then we would want.

2. a real life scenario: my package it used by a site that in a certain page the global `console` object is overridden by oracle portal (they use their own implementation of console, apparently..), the oracle version does implement the `warn` method - nor I or the site owners were aware of this and it ends up with an exception from Ajv that prevents some capabilities to work.
IMHO it is better to secure our code from such scenarios - if our code throws an exception due to bad input it is still our responsibility (unless it is part of the expected documented scenarios, of course..).

Anyhow - sorry for my insistence - at the end it is your call :)  if I still wasn't able to convince I will change the implementation as you suggested.

Reply:
Thank you. To address the problem above, whether with the passed object or with the default global console, rather than hiding the problem it is better to test the object has all 3 methods during the instance initialisation (in the constructor) and throw an exception if it doesn't:

```
if (!(logger.log && logger.warn && logger.error)) throw new Error('logger is invalid');
```

Reply:
this would not help with the scenario I described as at the initialization phase it was still the global console object, that was later changed by the oracle portal page.
My way to deal with it will be to disable the log once this change will be rolled out.

Reply:
It would, because once console is assigned to the instance the fact that oracle overridden a global wouldn't have affected it. It would still continue using the old one.

Reply:
correct :)
anyhow, it still throws an exception.. think of a package that uses a package that uses Ajv in an oracle page - quite a mess...

Reply:
The difference is that it would throw an exception early and always rather than late and sometimes (in case console is replaced before ajv initialisation; if it happens after as you wrote, it would simply work). I prefer this behaviour to simply not having warnings.

Reply:
Let's agree to disagree then :)
I will push a change soon.